enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,404 @@
+   node annotations, it will clear the `k8s.ovn.org/hybrid-overlay-xxx` node
+   annotation then set the annotation `k8s.ovn.org/node-subnets:` according to
+   the HostSubnet object of the node to bypass the ovnkube node subnet
+   allocation. 
+3. CNO will also update the hybrid overlay configuration and the onvkube-master
+   pods will take the updated hybrid overlay configuration then update the OVN
+   database.
+4. MCO will drain and reboot the node.
+5. After booting up, br-ex will be created and ovnkube-node will run in hybrid
+   overlay mode on the node.
+6. The wrapper script of ovnkube-node DaemonSet will see br-ex, then launch the
+   `ovnkube-node` process.
+7. MCO will uncordon the node. Pods will be created on the node using
+   ovn-kubernetes as the default CNI plugin.
+
+The above process will be repeated for each node, until all the nodes have been
+applied to the new MachineConfig and converted to OVN Kubernetes.
+
+#### Migration Cleanup
+
+Once migration is complete, CNO will:
+- delete the openshift-sdn DaemonSets
+- redeploy ovn-kubernetes in "normal" mode (no migration mode config, no node
+  affinity, hybrid overlay disabled).
+- remove the migration-related labels from the nodes
+
+### API
+
+A new field `spec.migration.isLive` will be introduced to the CRD
+`networks.operator.openshift.io`. Users can set this flag to conduct a live
+migration or a offline migration.
+
+To start the migration, users need to update the `network.operator`
+CR by adding:
+
+CNO will deploy both OpenShift SDN and OVN Kubernetes pods in 'migration' mode.
+
+```json
+{ 
+  "spec": { 
+    "migration": {
+      "networkType": "OVNKubernetes",
+      // isLive indicates whether we're doing a live migration or not.
+      "isLive": true
+    }
+  } 
+}
+```
+
+On removing of the `spec.migration` field, CNO will start the migration cleanup
+and redeploy the target CNI in 'normal' mode.
+
+CNO will also report the migration state to under the `status` of
+`network.operator` CR.
+
+```json
+{ 
+  "status": { 
+    "migration": {
+      // The state can be 'Setup', 'Working', 'Done' or 'Error', 
+      "state": "Working",
+      // the reason needs to be filled when state is 'Error'.
+      "reason": ""
+    }
+  } 
+}
+```
+
+### Rollback
+
+Users shall be able to rollback to openshift-sdn after the migration is
+complete. The migration is bidirectional, thus users can follow the similar
+above-mentioned procedure to conduct the rollback.
+
+### Lifecycle Management
+
+This is a one-time operation for a cluster, therefore no lifecycle management.
+
+### Test Plan
+
+We need to setup a CI job which can run the sig-network test against a cluster
+that is in an intermediate state of live migration, specifically some nodes are
+running with OVN Kubernetes, and the rests are running with OpenShift SDN. So
+that we can ensure the cluster network remains functioning during the migration.
```

Comment:
Depending on how long it takes to do a migration vs how long it takes to run the e2e suite, you could possibly have a job that just migrates the cluster back and forth until the e2e suite completes?

Reply:
Do we test the upgrade in this way?